Creating disposables not blocked when disposing all

This stack trace from Android application shows the problem:
```
Caused by: java.lang.IllegalArgumentException: Invalid shader handle: 0
                      at glimpse.shaders.ShaderHandle.<init>(Shader.kt:34)
                      at glimpse.android.gles.GLES.createShader(GLES.kt:66)
                      at glimpse.shaders.ProgramBuilder.invoke(ProgramBuilder.kt:16)
                      at glimpse.shaders.ProgramBuilder.vertexShader(ProgramBuilder.kt:30)
                      at glimpse.materials.TexturedShaderHelper$program$2$1.invoke(Textured.kt:50)
                      at glimpse.materials.TexturedShaderHelper$program$2$1.invoke(Textured.kt:46)
                      at glimpse.shaders.ProgramBuilderKt.shaderProgram(ProgramBuilder.kt:60)
                      at glimpse.materials.TexturedShaderHelper$program$2.invoke(Textured.kt:49)
                      at glimpse.materials.TexturedShaderHelper$program$2.invoke(Textured.kt:46)
                      at glimpse.gles.delegates.DisposableLazyDelegate.getValue(DisposableLazyDelegate.kt:21)
                      at glimpse.materials.TexturedShaderHelper.getProgram(Textured.kt)
                      at glimpse.materials.ShaderHelper.dispose(ShaderHelper.kt:159)
                      at glimpse.gles.Disposables.disposeAll(Disposable.kt:37)
                      at glimpse.android.example.tetrahedron.MainActivity.onDestroy(MainActivity.kt:119)
                      at android.app.Activity.performDestroy(Activity.java:6407)
                      at android.app.Instrumentation.callActivityOnDestroy(Instrumentation.java:1142)
                      at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:3818)
                      at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:3849) 
                      at android.app.ActivityThread.-wrap5(ActivityThread.java) 
                      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1398) 
                      at android.os.Handler.dispatchMessage(Handler.java:102) 
                      at android.os.Looper.loop(Looper.java:148) 
                      at android.app.ActivityThread.main(ActivityThread.java:5417) 
                      at java.lang.reflect.Method.invoke(Native Method) 
                      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
```

1. All disposables are being disposed.
2. A shader helper tries to delete its program.
3. Program field is not yet initialized (or already disposed), so the program is being created.
4. When creating a new shader, its handle is 0 (invalid).

--- a/api/src/main/kotlin/glimpse/gles/Disposable.kt
+++ b/api/src/main/kotlin/glimpse/gles/Disposable.kt
@@ -21,6 +21,8 @@ interface Disposable {
  */
 object Disposables {
 
+	internal var isDisposing = false
+
 	private val disposables = mutableSetOf<Disposable>()
 
 	/**
@@ -34,7 +36,9 @@ object Disposables {
 	 * Disposes all registered objects.
 	 */
 	fun disposeAll() {
+		isDisposing = true
 		disposables.forEach { it.dispose() }
 		disposables.clear()
+		isDisposing = false
 	}
 }

--- a/api/src/main/kotlin/glimpse/gles/delegates/DisposableLazyDelegate.kt
+++ b/api/src/main/kotlin/glimpse/gles/delegates/DisposableLazyDelegate.kt
@@ -1,6 +1,7 @@
 package glimpse.gles.delegates
 
 import glimpse.gles.Disposable
+import glimpse.gles.Disposables
 import kotlin.reflect.KProperty
 
 /**
@@ -9,19 +10,19 @@ import kotlin.reflect.KProperty
  * @param T Property type.
  * @property init Lazy initialization function.
  */
-class DisposableLazyDelegate<T>(private val init: () -> T) : Disposable {
+class DisposableLazyDelegate<T>(private val init: () -> T?) : Disposable {
 
 	private var value: T? = null
 
 	/**
 	 * Gets delegated property value.
 	 */
-	operator fun getValue(thisRef: Any?, property: KProperty<*>): T {
-		if (value == null) {
+	operator fun getValue(thisRef: Any?, property: KProperty<*>): T? {
+		if (value == null && !Disposables.isDisposing) {
 			value = init()
 			registerDisposable()
 		}
-		return value!!
+		return value
 	}
 
 	/**

--- a/api/src/main/kotlin/glimpse/materials/ShaderHelper.kt
+++ b/api/src/main/kotlin/glimpse/materials/ShaderHelper.kt
@@ -8,6 +8,7 @@ import glimpse.gles.*
 import glimpse.gles.delegates.GLESDelegate
 import glimpse.models.Mesh
 import glimpse.shaders.Program
+import glimpse.shaders.ProgramHandle
 import glimpse.textures.Texture
 import java.nio.FloatBuffer
 
@@ -24,19 +25,19 @@ abstract class ShaderHelper : Disposable {
 	/**
 	 * Shader program.
 	 */
-	protected abstract val program: Program
+	protected abstract val program: Program?
 
 	/**
 	 * Location of a shader attribute with the given name.
 	 */
 	protected val String.attributeLocation: AttributeLocation
-		get() = gles.getAttributeLocation(program.handle, this)
+		get() = gles.getAttributeLocation(program?.handle ?: ProgramHandle(0), this)
 
 	/**
 	 * Location of a shader uniform with the given name.
 	 */
 	protected val String.uniformLocation: UniformLocation
-		get() = gles.getUniformLocation(program.handle, this)
+		get() = gles.getUniformLocation(program?.handle ?: ProgramHandle(0), this)
 
 	/**
 	 * Name of a shader attribute containing vertex position.
@@ -57,7 +58,7 @@ abstract class ShaderHelper : Disposable {
 	 * Tells GLES implementation to use this program.
 	 */
 	fun use() {
-		program.use()
+		program?.use()
 	}
 
 	/**
@@ -156,6 +157,6 @@ abstract class ShaderHelper : Disposable {
 	 * Disposes shader helper.
 	 */
 	override fun dispose() {
-		program.delete()
+		program?.delete()
 	}
 }

--- a/api/src/test/kotlin/glimpse/gles/delegates/DisposableLazyDelegateSpec.kt
+++ b/api/src/test/kotlin/glimpse/gles/delegates/DisposableLazyDelegateSpec.kt
@@ -1,5 +1,6 @@
 package glimpse.gles.delegates
 
+import glimpse.gles.Disposable
 import glimpse.gles.Disposables
 import glimpse.test.GlimpseSpec
 
@@ -11,7 +12,7 @@ class DisposableLazyDelegateSpec : GlimpseSpec() {
 			"not change after initialization" {
 				var number = 1
 				class SomeClass {
-					val value : Int by DisposableLazyDelegate { number }
+					val value : Int? by DisposableLazyDelegate { number }
 				}
 				val instance = SomeClass()
 
@@ -25,7 +26,7 @@ class DisposableLazyDelegateSpec : GlimpseSpec() {
 			"change after it was disposed" {
 				var number = 1
 				class SomeClass {
-					val value : Int by DisposableLazyDelegate { number }
+					val value : Int? by DisposableLazyDelegate { number }
 				}
 				val instance = SomeClass()
 
@@ -40,8 +41,8 @@ class DisposableLazyDelegateSpec : GlimpseSpec() {
 			"not affect other disposable lazily delegated properties" {
 				var number = 1
 				class SomeClass {
-					val value1 : Int by DisposableLazyDelegate { number }
-					val value2 : Int by DisposableLazyDelegate { number }
+					val value1 : Int? by DisposableLazyDelegate { number }
+					val value2 : Int? by DisposableLazyDelegate { number }
 				}
 				val instance1 = SomeClass()
 				val instance2 = SomeClass()
@@ -68,6 +69,24 @@ class DisposableLazyDelegateSpec : GlimpseSpec() {
 				instance1.value2 shouldBe 2
 				instance2.value1 shouldBe 3
 				instance2.value2 shouldBe 4
+			}
+			"not be initialized when disposables are being disposed" {
+				class SomeClass : Disposable {
+
+					init {
+						registerDisposable()
+					}
+
+					val value : Int? by DisposableLazyDelegate { throw AssertionError("Value is being initialized") }
+
+					override fun dispose() {
+						value?.inc()
+					}
+				}
+
+				val instance = SomeClass()
+
+				Disposables.disposeAll()
 			}
 		}
 

--- a/materials/src/main/kotlin/glimpse/materials/Plastic.kt
+++ b/materials/src/main/kotlin/glimpse/materials/Plastic.kt
@@ -39,7 +39,7 @@ class Plastic(val diffuse: Color, val ambient: Color = diffuse, val specular: Co
 
 internal object PlasticShaderHelper : ShaderHelper() {
 
-	override val program: Program by DisposableLazyDelegate {
+	override val program: Program? by DisposableLazyDelegate {
 		shaderProgram {
 			vertexShader {
 				PlasticShaderHelper.resource("Plastic_vertex.glsl").lines.joinToString(separator = "\n") { it }

--- a/materials/src/main/kotlin/glimpse/materials/Textured.kt
+++ b/materials/src/main/kotlin/glimpse/materials/Textured.kt
@@ -45,7 +45,7 @@ class Textured(val shininess: Float = 100f, val texture: (TextureType) -> Textur
 
 internal object TexturedShaderHelper : ShaderHelper() {
 
-	override val program: Program by DisposableLazyDelegate {
+	override val program: Program? by DisposableLazyDelegate {
 		shaderProgram {
 			vertexShader {
 				TexturedShaderHelper.resource("Textured_vertex.glsl").lines.joinToString(separator = "\n") { it }

